File: internal/updater/updater.go
package updater

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/minio/selfupdate"
)

const repo = "icepie/rdev"

var defaultProxyPrefixes = []string{
	"",
	"https://gh-proxy.com/",
	"https://gh-proxy.net/",
	"https://gh.llkk.cc/",
	"https://hub.gitmirror.com/",
}

type Config struct {
	App      string
	Version  string
	Enabled  bool
	Interval time.Duration
	Logger   *log.Logger
}

type releaseInfo struct {
	TagName    string         `json:"tag_name"`
	Draft      bool           `json:"draft"`
	Prerelease bool           `json:"prerelease"`
	Assets     []releaseAsset `json:"assets"`
}

type releaseAsset struct {
	Name string `json:"name"`
	URL  string `json:"browser_download_url"`
}

func Start(ctx context.Context, cfg Config) {
	if !cfg.Enabled {
		return
	}
	if cfg.Interval <= 0 {
		cfg.Interval = time.Minute
	}
	if cfg.Logger == nil {
		cfg.Logger = log.Default()
	}
	go run(ctx, cfg)
}

func run(ctx context.Context, cfg Config) {
	timer := time.NewTimer(time.Minute)
	defer timer.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-timer.C:
			updated, err := CheckAndApply(ctx, cfg)
			if err != nil {
				cfg.Logger.Printf("auto-update check failed: %v", err)
			} else if updated {
				cfg.Logger.Printf("auto-update applied; restarting into new version")
				if err := restartSelf(); err != nil {
					cfg.Logger.Printf("auto-update restart failed: %v", err)
				}
				return
			}
			timer.Reset(cfg.Interval)
		}
	}
}

func CheckAndApply(ctx context.Context, cfg Config) (bool, error) {
	current := normalizeVersion(cfg.Version)
	if current == "" || current == "dev" {
		return false, nil
	}
	release, err := latestRelease(ctx)
	if err != nil {
		return false, err
	}
	latest := normalizeVersion(release.TagName)
	if latest == "" || !newerVersion(latest, current) {
		return false, nil
	}
	assetName := releaseAssetName(cfg.App)
	assetURL := ""
	for _, asset := range release.Assets {
		if asset.Name == assetName {
			assetURL = asset.URL
			break
		}
	}
	if assetURL == "" {
		return false, fmt.Errorf("release %s asset %s not found", release.TagName, assetName)
	}
	data, err := downloadWithProxies(ctx, assetURL)
	if err != nil {
		return false, err
	}
	if err := selfupdate.Apply(bytes.NewReader(data), selfupdate.Options{}); err != nil {
		if rollbackErr := selfupdate.RollbackError(err); rollbackErr != nil {
			return false, fmt.Errorf("apply update failed: %w; rollback failed: %v", err, rollbackErr)
		}
		return false, err
	}
	return true, nil
}

func latestRelease(ctx context.Context) (*releaseInfo, error) {
	url := "https://api.github.com/repos/" + repo + "/releases/latest"
	data, err := downloadWithProxies(ctx, url)
	if err != nil {
		return nil, err
	}
	var release releaseInfo
	if err := json.Unmarshal(data, &release); err != nil {
		return nil, err
	}
	if release.TagName == "" || release.Draft || release.Prerelease {
		return nil, errors.New("no stable latest release")
	}
	return &release, nil
}

func downloadWithProxies(ctx context.Context, target string) ([]byte, error) {
	client := &http.Client{Timeout: 3 * time.Minute}
	var lastErr error
	for _, prefix := range proxyPrefixes() {
		url := proxiedURL(prefix, target)
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		if err != nil {
			lastErr = err
			continue
		}
		req.Header.Set("User-Agent", "rdev-auto-updater")
		resp, err := client.Do(req)
		if err != nil {
			lastErr = err
			continue
		}
		body, readErr := io.ReadAll(resp.Body)
		resp.Body.Close()
		if readErr != nil {
			lastErr = readErr
			continue
		}
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			lastErr = fmt.Errorf("%s returned %s", url, resp.Status)
			continue
		}
		return body, nil
	}
	if lastErr == nil {
		lastErr = errors.New("all update URLs failed")
	}
	return nil, lastErr
}

func proxyPrefixes() []string {
	prefixes := []string{}
	if env := strings.TrimSpace(os.Getenv("RDEV_UPDATE_PROXY")); env != "" {
		for _, part := range strings.Split(env, ",") {
			if p := strings.TrimSpace(part); p != "" {
				prefixes = append(prefixes, p)
			}
		}
	}
	prefixes = append(prefixes, defaultProxyPrefixes...)
	seen := map[string]bool{}
	out := make([]string, 0, len(prefixes))
	for _, prefix := range prefixes {
		if seen[prefix] {
			continue
		}
		seen[prefix] = true
		out = append(out, prefix)
	}
	return out
}

func proxiedURL(prefix, target string) string {
	if prefix == "" {
		return target
	}
	if strings.Contains(prefix, "${url}") {
		return strings.ReplaceAll(prefix, "${url}", target)
	}
	return strings.TrimRight(prefix, "/") + "/" + target
}

func releaseAssetName(app string) string {
	ext := ""
	if runtime.GOOS == "windows" {
		ext = ".exe"
	}
	return fmt.Sprintf("rdev-%s-%s-%s%s", app, runtime.GOOS, runtime.GOARCH, ext)
}

func normalizeVersion(version string) string {
	version = strings.TrimSpace(version)
	version = strings.TrimPrefix(version, "v")
	if idx := strings.IndexAny(version, "-+"); idx >= 0 {
		version = version[:idx]
	}
	if version == "" || version[0] < '0' || version[0] > '9' {
		return "dev"
	}
	return version
}

func newerVersion(latest, current string) bool {
	la := versionParts(latest)
	cu := versionParts(current)
	for i := 0; i < 3; i++ {
		if la[i] != cu[i] {
			return la[i] > cu[i]
		}
	}
	return false
}

func versionParts(version string) [3]int {
	var out [3]int
	parts := strings.Split(version, ".")
	for i := 0; i < len(parts) && i < 3; i++ {
		fmt.Sscanf(parts[i], "%d", &out[i])
	}
	return out
}

File: clients/rdev-client-gpu/src/updater.rs
use std::{
    env,
    fs::File,
    path::{Path, PathBuf},
    thread,
    time::{Duration, Instant},
};

use anyhow::{anyhow, Context, Result};
use self_update::{backends::github, Download, Extract};
use tempfile::TempDir;
use tracing::{info, warn};

const REPO_OWNER: &str = "icepie";
const REPO_NAME: &str = "rdev";

const DEFAULT_PROXY_PREFIXES: &[&str] = &[
    "",
    "https://gh-proxy.com/",
    "https://gh-proxy.net/",
    "https://gh.llkk.cc/",
    "https://hub.gitmirror.com/",
];

#[derive(Debug, Clone)]
pub struct Config {
    pub version: &'static str,
    pub enabled: bool,
    pub interval: Duration,
}

pub fn start(cfg: Config) {
    if !cfg.enabled {
        return;
    }
    let interval = if cfg.interval.is_zero() {
        Duration::from_secs(60)
    } else {
        cfg.interval
    };
    thread::spawn(move || run(Config { interval, ..cfg }));
}

fn run(cfg: Config) {
    let mut failures = FailureLog::default();
    thread::sleep(Duration::from_secs(60));
    loop {
        match check_and_apply(&cfg) {
            Ok(true) => {
                info!("auto-update applied; restart rdev-client-gpu to use the new version");
                return;
            }
            Ok(false) => failures.recovered(),
            Err(err) => failures.record(&err),
        }
        thread::sleep(cfg.interval);
    }
}

pub fn check_and_apply(cfg: &Config) -> Result<bool> {
    let current = normalize_version(cfg.version);
    if current.is_empty() || current == "dev" {
        return Ok(false);
    }

    let updater = github::Update::configure()
        .repo_owner(REPO_OWNER)
        .repo_name(REPO_NAME)
        .current_version(&current)
        .bin_name(binary_name())
        .show_output(false)
        .build()
        .context("build GitHub updater")?;

    let release = updater
        .get_latest_release()
        .context("fetch latest GitHub release")?;
    let Some(release) = release.latest() else {
        return Ok(false);
    };

    let latest = normalize_version(&release.version);
    if latest.is_empty() || !newer_version(&latest, &current) {
        return Ok(false);
    }

    let asset_name = release_asset_name().ok_or_else(|| {
        anyhow!(
            "auto-update is not supported for {}/{}",
            env::consts::OS,
            env::consts::ARCH
        )
    })?;
    let asset = release
        .assets
        .iter()
        .find(|asset| asset.name == asset_name)
        .ok_or_else(|| {
            anyhow!(
                "release v{} asset {} not found",
                release.version,
                asset_name
            )
        })?;

    let tmp = TempDir::new().context("create updater temp dir")?;
    let archive_path = tmp.path().join(&asset.name);
    let download_url = github_download_url(&release.version, &asset.name);
    download_with_proxies(&download_url, &archive_path)
        .with_context(|| format!("download release asset {}", asset.name))?;

    let bin_path = PathBuf::from(package_dir_name(&asset.name)).join(binary_name());
    Extract::from_source(&archive_path)
        .extract_file(tmp.path(), &bin_path)
        .with_context(|| format!("extract {} from {}", bin_path.display(), asset.name))?;

    let new_exe = tmp.path().join(&bin_path);
    self_replace::self_replace(&new_exe).context("replace current executable")?;

    Ok(true)
}

fn download_with_proxies(target: &str, archive_path: &Path) -> Result<()> {
    let mut last_err = None;
    for prefix in proxy_prefixes() {
        let url = proxied_url(&prefix, target);
        let mut file = File::create(archive_path)
            .with_context(|| format!("create {}", archive_path.display()))?;
        let mut download = Download::from_url(&url);
        let _ = download.header("User-Agent", "rdev-auto-updater");
        let _ = download.header("Accept", "application/octet-stream");
        if let Err(err) = download.download_to(&mut file) {
            last_err = Some(err);
            continue;
        }
        return Ok(());
    }
    Err(anyhow!(
        "{}",
        last_err
            .map(|err| err.to_string())
            .unwrap_or_else(|| "all update URLs failed".to_string())
    ))
}

fn binary_name() -> &'static str {
    if cfg!(windows) {
        "rdev-client-gpu.exe"
    } else {
        "rdev-client-gpu"
    }
}

fn package_dir_name(asset_name: &str) -> &str {
    asset_name
        .strip_suffix(".tar.gz")
        .or_else(|| asset_name.strip_suffix(".zip"))
        .unwrap_or(asset_name)
}

fn github_download_url(version: &str, asset_name: &str) -> String {
    format!(
        "https://github.com/{REPO_OWNER}/{REPO_NAME}/releases/download/v{}/{}",
        normalize_version(version),
        asset_name
    )
}

fn release_asset_name() -> Option<String> {
    let arch = match env::consts::ARCH {
        "x86_64" => "amd64",
        "aarch64" => "arm64",
        "arm" => "armv7",
        "x86" => "x86",
        other => other,
    };
    match env::consts::OS {
        "windows" => {
            #[cfg(windows)]
            if crate::winpty::is_legacy_windows() {
                return Some("rdev-client-gpu-windows-win7-amd64.zip".to_string());
            }
            Some(format!("rdev-client-gpu-windows-{arch}.zip"))
        }
        "linux" | "macos" | "android" => Some(format!(
            "rdev-client-gpu-{}-{arch}.tar.gz",
            if env::consts::OS == "macos" {
                "darwin"
            } else {
                env::consts::OS
            }
        )),
        _ => None,
    }
}

fn normalize_version(version: &str) -> String {
    let mut version = version.trim().trim_start_matches('v');
    if let Some(idx) = version.find(['-', '+']) {
        version = &version[..idx];
    }
    if version.is_empty() || !version.as_bytes()[0].is_ascii_digit() {
        return "dev".to_string();
    }
    version.to_string()
}

fn newer_version(latest: &str, current: &str) -> bool {
    let latest = version_parts(latest);
    let current = version_parts(current);
    latest > current
}

fn version_parts(version: &str) -> [u64; 3] {
    let mut out = [0; 3];
    for (idx, part) in version.split('.').take(3).enumerate() {
        out[idx] = part.parse::<u64>().unwrap_or(0);
    }
    out
}

fn proxy_prefixes() -> Vec<String> {
    let mut prefixes = Vec::new();
    if let Ok(env) = env::var("RDEV_UPDATE_PROXY") {
        prefixes.extend(
            env.split(',')
                .map(str::trim)
                .filter(|prefix| !prefix.is_empty())
                .map(ToOwned::to_owned),
        );
    }
    prefixes.extend(
        DEFAULT_PROXY_PREFIXES
            .iter()
            .map(|prefix| (*prefix).to_string()),
    );

    let mut out = Vec::with_capacity(prefixes.len());
    for prefix in prefixes {
        if !out.contains(&prefix) {
            out.push(prefix);
        }
    }
    out
}

fn proxied_url(prefix: &str, target: &str) -> String {
    if prefix.is_empty() {
        return target.to_string();
    }
    if prefix.contains("${url}") {
        return prefix.replace("${url}", target);
    }
    format!("{}/{}", prefix.trim_end_matches('/'), target)
}

#[derive(Debug)]
struct FailureLog {
    last_err: String,
    suppressed: usize,
    last_logged_at: Instant,
}

impl Default for FailureLog {
    fn default() -> Self {
        Self {
            last_err: String::new(),
            suppressed: 0,
            last_logged_at: Instant::now() - Duration::from_secs(15 * 60),
        }
    }
}

impl FailureLog {
    fn record(&mut self, err: &anyhow::Error) {
        let message = err.to_string();
        if message != self.last_err {
            if !self.last_err.is_empty() && self.suppressed > 0 {
                warn!(
                    "auto-update check changed after suppressing {} repeated failures",
                    self.suppressed
                );
            }
            self.last_err = message;
            self.suppressed = 0;
            self.last_logged_at = Instant::now();
            warn!("auto-update check failed: {err:#}");
            return;
        }

        self.suppressed += 1;
        if self.last_logged_at.elapsed() >= Duration::from_secs(15 * 60) {
            warn!(
                "auto-update check still failing; suppressed {} repeated failures: {err:#}",
                self.suppressed
            );
            self.suppressed = 0;
            self.last_logged_at = Instant::now();
        }
    }

    fn recovered(&mut self) {
        if self.last_err.is_empty() {
            return;
        }
        if self.suppressed > 0 {
            info!(
                "auto-update check recovered; suppressed {} repeated failures",
                self.suppressed
            );
        }
        self.last_err.clear();
        self.suppressed = 0;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_versions() {
        assert_eq!(normalize_version("v0.2.98"), "0.2.98");
        assert_eq!(normalize_version("0.2.98-dirty"), "0.2.98");
        assert_eq!(normalize_version("main"), "dev");
    }

    #[test]
    fn compares_versions() {
        assert!(newer_version("0.2.99", "0.2.98"));
        assert!(newer_version("0.3.0", "0.2.99"));
        assert!(!newer_version("0.2.98", "0.2.98"));
        assert!(!newer_version("0.2.97", "0.2.98"));
    }

    #[test]
    fn strips_package_suffixes() {
        assert_eq!(
            package_dir_name("rdev-client-gpu-linux-amd64.tar.gz"),
            "rdev-client-gpu-linux-amd64"
        );
        assert_eq!(
            package_dir_name("rdev-client-gpu-windows-amd64.zip"),
            "rdev-client-gpu-windows-amd64"
        );
    }

    #[test]
    fn builds_proxy_urls() {
        let target = "https://github.com/icepie/rdev/releases/download/v0.2.98/a.zip";
        assert_eq!(proxied_url("", target), target);
        assert_eq!(
            proxied_url("https://gh-proxy.com/", target),
            format!("https://gh-proxy.com/{target}")
        );
        assert_eq!(
            proxied_url("http://proxy/?url=${url}", target),
            format!("http://proxy/?url={target}")
        );
    }

    #[test]
    fn builds_github_download_url() {
        assert_eq!(
            github_download_url("v0.2.98", "rdev-client-gpu-linux-amd64.tar.gz"),
            "https://github.com/icepie/rdev/releases/download/v0.2.98/rdev-client-gpu-linux-amd64.tar.gz"
        );
    }
}
